### DISCUSSION TITLE
Setting up development environment for Freeplane. Help needed!

### DISCUSSION THREAD

**Discussion by EdoFro** [freeplane/freeplane#1321] [Help] [answered]
Hi,
I'm trying to set up the development environment for Freeplane but I'm failing again. I have tried many times but I am unable to overcome the different problems and end up confused and frustrated.

It is clear that I am not an expert on the subject and that I am probably the root of the problem, but I still feel that installing the environment should be something simple and automatic and not kind of black magic.

I am going to describe the steps I have done and the errors that appear and hopefully there is someone who can guide me.

## the facts

Operating System:

```
Windows 11 Pro
Version	22H2
Instaled‎   02-‎05-‎2023
OS version	22621.1992
Windows Feature Experience Pack 1000.22644.1000.0
```

1. I installed intelliJ as my IDE a few months ago:

```
IntelliJ IDEA 2023.1 (Community Edition)
Build #IC-231.8109.175, built on March 28, 2023
Runtime version: 17.0.6+10-b829.5 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 11.0
GC: G1 Young Generation, G1 Old Generation
Memory: 2024M
Cores: 8
```

2. I cloned Freeplane to my Github Repo, then to my local drive and I created a new branch.

3. I tried to follow the [IDE setup documentation in Freeplane Docs](https://docs.freeplane.org/coding/IDE_setup.html): more specifically the steps in [Open the freeplane checkout (IntelliJ 2021)](https://docs.freeplane.org/coding/IDE_setup.html#open-the-freeplane-checkout-intellij-2021)

In the second step: ***"Point to the build.gradle in the checkout and click OK, then Open as Project"***:
- I selected the build.gradle file in the root directory
- clicked ok
- and clicked "Open as a Project"

Then IntelliJ started opening freeplane as a project, opened the README.md file and the **Build** window doing some commands.

And it ended this way:
![image](https://github.com/freeplane/freeplane/assets/35700575/25ba6bc2-0425-4878-bb7a-b2526f70ab01)

Any ideas??
@macmarrum? @dpolivaev?

Thanks in advance,
edo

**Comment by EdoFro** [comment]
Gradle version:
```
C:\Users\efroh>gradle -v

------------------------------------------------------------
Gradle 8.1.1
------------------------------------------------------------

Build time:   2023-04-21 12:31:26 UTC
Revision:     1cf537a851c635c364a4214885f8b9798051175b

Kotlin:       1.8.10
Groovy:       3.0.15
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          17.0.6 (Eclipse Adoptium 17.0.6+10)
OS:           Windows 11 10.0 amd64

```

**Comment by EdoFro** [comment]
> 4\. [2021)](https://docs.freeplane.org/coding/IDE_setup.html#open-the-freeplane-checkout-intellij-2021)
> 
> In the second step: _**"Point to the build.gradle in the checkout and click OK, then Open as Project"**_:

I don't know if I selected the right **build.gradle** file.
There are 16

![image](https://github.com/freeplane/freeplane/assets/35700575/ffb832dd-e4f6-447c-a203-e94032e04eaf)

**Comment by macmarrum** [comment, marked as answer]
Hi @EdoFro 

I am not sure the page https://docs.freeplane.org/coding/IDE_setup.html is still valid.\
Modern IntelliJ IDEA does all the Gradle-importing work for you. The overall process is simplified nowadays.

Here's how I set up Freeplane in IntelliJ IDEA

1. `File->New->Project from Existing Sources` and point to your cloned repo
2. In "Import Project" window that appears, choose "Import project from external model" -> "Gradle"
3. "Trust Project"

This should import and run Gradle for the project, downloading all dependencies.

To build the freeplane binaries, open Terminal (at the bottom, in IDEA) and run
```powershell
.\gradlew.bat build
```

If it complains about JAVA_HOME not being set, you need to set the variable, preferably for the system, or for this session only:
```powershell
$Env:JAVA_HOME="C:\Users\<user-name>\.jdks\temurin-<version>"
.\gradlew.bat build
```

Note: IDEA has a Gradle sidebar, on the right-hand side. You can try running the build from there
* freeplane_root
  * Tasks
    * build
      * build

It doesn't work for me, because IDEA complains about missing imports in the class `org.freeplane.plugin.script.FreeplaneScriptBaseClass`.

Once it's built, you can launch Freeplane from the Terminal in IDEA
EDIT: <del>.\BIN\freeplaneConsole.exe</del>
```powershell
cd BIN
.\freeplane.bat
```

If it complains about 'java' not being recognized as an internal or external command, you need to expand PATH by adding your java.exe-containing directory, preferably for the system, or for this session only:

```powershell
$Env:PATH="${Env:PATH};C:\Users\<user-name>\.jdks\temurin-<version>\bin"
.\freeplane.bat
```

**Reply by EdoFro** [reply]
Thanks @macmarrum!
I followed your instructions and it worked fine!!


About this:
> Note: IDEA has a Gradle sidebar, on the right-hand side. You can try running the build from there
> 
> * freeplane_root
>   * Tasks
>     * build
>       * build
> 
> It doesn't work for me, because IDEA complains about missing imports in the class `org.freeplane.plugin.script.FreeplaneScriptBaseClass`.

It didn't work. I got the same message.

**Reply by euu2021** [reply]
I'm trying to build using Gradle, but it doesn't generate exe files. It generates the BIN folder, but it doesn't have the exe files:

![image](https://github.com/freeplane/freeplane/assets/77707706/4f22cb0a-426d-431a-8009-d7aa657ef274)


> It doesn't work for me, because IDEA complains about missing imports in the class org.freeplane.plugin.script.FreeplaneScriptBaseClass.

Do you mean, you get those errors when trying to build from the sidebar, but not when building from the terminal? I get those errors even when building from the terminal.

**Reply by dpolivaev** [reply]
You need launch4j installed and LAUNCH4J_DIR env variable set.
I use my custom version of launch4j, see https://github.com/dpolivaev/launch4j

**Reply by euu2021** [reply]
> You need launch4j installed and LAUNCH4J_DIR env variable set.
I use my custom version of launch4j, see https://github.com/dpolivaev/launch4j

It worked. Thanks

**Reply by macmarrum** [reply]
I've updated the instructions to use `freeplane.bat` instead of `freeplaneConsole.exe`, so that it's also possible to go without [launch4j](https://github.com/dpolivaev/launch4j).

**Reply by euu2021** [reply]
To test, I disabled launch4j and started a project from scratch. But, when I try to run freeplane.bat from the terminal, it doesn't work, and gives this message: 

> C:\Users\kauen\Documents\ferramentas\freeplane\FPsource\test2\freeplane-1.11.x\BIN>java --add-exports=java.desktop/sun.awt.shell=ALL-UNNAMED --add-exports=java.desktop/com.sun.java.swing.plaf.windows=ALL-UNNAMED -Xmx1g -Dorg.freeplane.userfpdir="C:\Users\kauen\AppData\Roaming\\Freeplane" -jar freeplanelauncher.jar
Exception in thread "main" java.lang.UnsupportedOperationException: The Security Manager is deprecated and will be removed in a future release
        at java.base/java.lang.System.setSecurityManager(System.java:431)
        at org.freeplane.launcher.Launcher.launchWithoutUICheck(Launcher.java:291)
        at org.freeplane.launcher.Launcher.main(Launcher.java:88)

This also fails: .\freeplane.bat -Xlint:-deprecation

I tried with multiple JDKs and Gradle JVMs.

But, I can run freeplane.bat using the run configs that are inside freeplane-1.11.x\freeplane_debughelper\idea\runConfigurations, as explained at https://docs.freeplane.org/attic/old-mediawiki-content/Debugging_scripts.html

**Reply by syeats-icm** [reply]
> Hi @EdoFro
> 
> I am not sure the page https://docs.freeplane.org/coding/IDE_setup.html is still valid. Modern IntelliJ IDEA does all the Gradle-importing work for you. The overall process is simplified nowadays.
> 
> Here's how I set up Freeplane in IntelliJ IDEA
> 
> 1. `File->New->Project from Existing Sources` and point to your cloned repo
> 2. In "Import Project" window that appears, choose "Import project from external model" -> "Gradle"
> 3. "Trust Project"
> 
> This should import and run Gradle for the project, downloading all dependencies.
> 
> To build the freeplane binaries, open Terminal (at the bottom, in IDEA) and run
> 
> ```powershell
> .\gradlew.bat build
> ```
> 
> If it complains about JAVA_HOME not being set, you need to set the variable, preferably for the system, or for this session only:
> 
> ```powershell
> $Env:JAVA_HOME="C:\Users\<user-name>\.jdks\temurin-<version>"
> .\gradlew.bat build
> ```
> 
--
> 
> Once it's built, you can launch Freeplane from the Terminal in IDEA EDIT: ~.\BIN\freeplaneConsole.exe~
> 
> ```powershell
> cd BIN
> .\freeplane.bat
> ```
--
> ```

This procedure worked for me, perfect, many thanks @macmarrum .

**Comment by dpolivaev** [comment]
I've checked it. The problem stemmed from using Gradle 8 because Freeplane dependencies hadn't been updated to support Gradle 8. I used Gradle 7.5.1 myself. However, I've now fixed it on branch 1.11.x and for Freeplane version 1.11.6. So, if you follow the instructions for IntelliJ, it should work fine now.

There's only one undocumented detail: for running/debugging Freeplane on MacOS, you should use Java 15 instead of Java 8. However, this issue isn't relevant for Windows.

**Reply by EdoFro** [reply]
Thanks @dpolivaev,
I updated my fork and then applied the steps mentioned by @macmarrum and it worked flawlessly.

**Comment by sifran-github** [comment]
I have a few questions regarding a "developer environment":
1. What do you mean exactly by saying a "Setting up development environment for Freeplane"? is it for autocomplete of code? what else? What I mean is: What is the difference between setting up a "developer environment" like that, and just downloading the source code from github and opening it in some editor?
2. What is this "developer environment" used for? is it for developing Freeplane itself for pushing code to main repository? developing add-ons? both? other things?
3. Can you completely compile Freeplane from source? or it requires additional knowledge / tools / processes?
4. Does anybody setup VS Code as the development environment?, or knows how to set it up

**Reply by nnako** [reply]
Hi @sifran-github ,

let me answer your questions and add a couple of comments...

> What do you mean exactly by saying a "Setting up development environment for Freeplane"? is it for autocomplete of code? what else? What I mean is: What is the difference between setting up a "developer environment" like that, and just downloading the source code from github and opening it in some editor?

The term "development environment" is widely used to describe a code editor in combination with further built-in or externally provided tools for

- editing of code (e.g. autocompletion, syntax highlighting, auto-indentation, ...)
- debugging (using breakpoints, variable watches, memory watches, stack tracing, ...) 
- version management (e.g. connecting to Freeplane's GitHub repo)
- building (providing connections to compilers, ...)
- ...

of / for an application from its source code. There are many different development environments out there, and it seems that just a few of them (e.g. VS code, Eclipse or VIM) are aimed to be used with different programming languages (not only Java). And here comes the difficulty: Every development environment has its own concept for providing user access to its features. e.g.:

- concerning the user interface, e.g. keyboard shortcuts are not unique but have to be studied and learned from scratch as soon as the development environment is changed.
- concerning the  configuration, e.g. where and how to define dependencies to external tools is to be figured out, individually
- ...

As you mentioned, @sifran-github, a simple text editor would do the basic job. But this practice might miss out on all the helpful features which a development environment provides out of the box.

> What is this "developer environment" used for? is it for developing Freeplane itself for pushing code to main repository? developing add-ons? both? other things?

everything you mentioned, except for Groovy script development. You could use this development environment for everything related to programming for Freeplane in Java language. For Groovy scripting, there might be a different "development environment" to be installed and used. In the best case, switching to Groovy development would be just re-configuration within the current development environment. 

> Can you completely compile Freeplane from source? or it requires additional knowledge / tools / processes?

You can build your own Freeplane executable. It will automatically appear in your local output folder.

> Does anybody setup VS Code as the development environment?, or knows how to set it up

;-) that is one of the problems. each development environment setup needs its own setup documentation. If you don't find a setup description for your preferred development environment here, you could be the first one creating (and frequently updating) it.

Hope, this helps understanding the complexity of this issue a bit. As each dev has his own (and slightly different) view on the use of development environments, I hope not to have offended anybody by my own view.

**Reply by sifran-github** [reply]
Thanks for the answer.

> * editing of code (e.g. autocompletion, syntax highlighting, auto-indentation, ...)
> * debugging (using breakpoints, variable watches, memory watches, stack tracing, ...)
> * version management (e.g. connecting to Freeplane's GitHub repo)
> * building (providing connections to compilers, ...)

Are all of these possible with Freeplane's codebase after setting up the dev environment?
I would like to setup such environment for myself, if its possible to fully compile Freeplane and also run it with breakpoints. Maybe I'll be able to understand the codebase better.
If its possible, I would try it. I prefer VS Code but if I won't be able to install it there, I will try intelliJ.

**Reply by euu2021** [reply]
> Maybe I'll be able to understand the codebase better.

Please, somehow share your findings. It will be helpful when other people try to understand the codebase.

**Comment by euu2021** [comment]
For future reference, those are the steps that worked for me:

Follow those steps: https://github.com/freeplane/freeplane/discussions/1321#discussioncomment-6631241, but you may not be able to complete the last step. 

A BIN folder will be created inside the root folder. If there are no exe files inside the BIN folder, then follow those steps: https://github.com/freeplane/freeplane/discussions/1321#discussioncomment-8952036. To set the variable LAUNCH4J_DIR, follow those steps:



<details>
  <summary>Steps:</summary>
  
Step 1: Download and Install Launch4j
Download Launch4j from the official website: Launch4j.
Extract the downloaded file to a folder on your computer, for example, C:\Launch4j.
Step 2: Set Up the LAUNCH4J_DIR Environment Variable
Now, you need to let your system know where Launch4j is installed by setting up an environment variable named LAUNCH4J_DIR.

Press Windows + R keys to open the Run dialog box.
Type sysdm.cpl and press Enter. This will open the System Properties.
Go to the "Advanced" tab and click on "Environment Variables".
In the "System variables" section, click on "New..." to create a new system variable.
In the "Variable name" field, enter LAUNCH4J_DIR.
In the "Variable value" field, enter the path where you extracted Launch4j, for example, C:\Launch4j.
Click OK to close the windows and save the variable.
Step 3: Verify the Configuration
To verify that the environment variable was set up correctly, you can open the Command Prompt and type the following command:
echo %LAUNCH4J_DIR%
If everything is configured correctly, the command should display the path you set for the LAUNCH4J_DIR variable.
  
</details>

Restart IDEA and rebuild.

Now, try to run FP: In the console, make sure the root folder is selected, then use: .\BIN\freeplaneConsole.exe

If it works, now it's time to prepare the run settings, so you can run FP in debug mode. The basic idea is in [the documentation](https://docs.freeplane.org/attic/old-mediawiki-content/Debugging_scripts.html), although it's written for Eclipse. For me, it complained about Gradle 9 or something like that, and I had to go into the "run configuration" of the "freeplane run" and change the path of the JRE. I changed to the Java 8 installation folder in my PC.

Now, you will be able to run on debug mode.

This "run configuration" is also the place to pass parameter to the exe, like pointing to a different user folder.  Writing like this creates a new APPDATA folder inside the BIN folder: -U "%APPDATA%\FreeplaneTests"


Inside the root folder, there is a file called freeplane_plugin_script\scripts. The FP launched from IDEA will load any scripts that are at this folder. The rest is explained at: https://docs.freeplane.org/attic/old-mediawiki-content/Debugging_scripts.html

**Comment by dpolivaev** [comment]
It looks like there might be confusion.

All launchers like any exe launchers, freeplane.sh and freeplane.bat have the same purpose namely starting java with correct parameters. For debugging, I would use IDE launch configurations and no other launchers at all. Also, there is no need to build exe files, they do not change so often and in most cases, you can use exe from Freeplane distribution, they are very likely the same as the ones you build yourself.

**Reply by macmarrum** [reply]
To expand on Dimitry's comment, I can see `run` and `runDist` configurations in `freeplane_debughelper/build.gradle`. 

The file can be edited to add options like `"-Dorg.freeplane.userfpdir=${System.env.APPDATA}/Freeplane"`.

Once freeplane is built (`.\gradlew.bat build`), `runDist` can be launched in a debug mode (from the right-click context menu).

> Inside the root folder, there is a file called freeplane_plugin_script\scripts. The FP launched from IDEA will load any scripts that are at this folder. The rest is explained at: https://docs.freeplane.org/attic/old-mediawiki-content/Debugging_scripts.html

`runDist` runs freeplane from the BIN directory, so scripts are in `BIN/scripts`

![2024-03-30_13,53,39](https://github.com/freeplane/freeplane/assets/7197131/2853c2d6-545a-4831-8b20-3ceb9b0e6439)

![2024-03-30_14,09,51](https://github.com/freeplane/freeplane/assets/7197131/8e9f0833-812e-42bb-9dc8-4faea93c2945)

**Reply by euu2021** [reply]
Now, I see. When I said `But, I can run freeplane.bat using the run configs that are inside freeplane-1.11.x\freeplane_debughelper\idea\runConfigurations` I thought I was using the freeplane.bat launcher, but, actually, I was using the IDE launch. So, the run and runDist configurations in freeplane_debughelper are launchers, themselves.
 
 > Once freeplane is built (.\gradlew.bat build), runDist can be launched in a debug mode (from the right-click context menu).
 
 For me, executing directly from the right-click context menu doesn't work. It only worked when I pasted into the .idea folder, like explained in the documentation.
 
There are multiple generations of guides in the documentation and in the forums, for different parts of the setup. I will write a page for the documentation, to be a guide to navigate through these other guides.

**Reply by dpolivaev** [reply]
It would be nice if someone would move relevant pages from "attic" to the main page set.

**Reply by euu2021** [reply]
> It would be nice if someone would move relevant pages from "attic" to the main page set.

Ok, I will try to organize the pages related to Scripting.

**Comment by euu2021** [comment]
Now, I noticed that, in my comments above, I was linking to the wrong page when I was talking about runConfigurations. This is the section that explain about it: https://docs.freeplane.org/coding/IDE_setup.html?highlight=runconfi#open-the-freeplane-checkout-intellij-2021

Another thing that I noticed, and I will include in the documentation, is that If I build with Gradle JVM 22, I get this message, even if I try to run with JDK 22:

```
Unsupported Java. 
Your build is currently configured to use Java 22 and Gradle 8.4.

Possible solution:
 - Use Java 20 as Gradle JVM: Open Gradle settings 
 - Upgrade Gradle wrapper to 4.5 version and re-import the project
```
 
 Gradle JVM 18 works fine.
